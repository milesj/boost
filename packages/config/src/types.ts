import { ModuleResolver, PackageStructure, Path } from '@boost/common';
import { DeepPartial } from '@boost/common/optimal';

export type ExtType = 'cjs' | 'js' | 'json' | 'json5' | 'mjs' | 'ts' | 'yaml' | 'yml';

export type LoaderType = Exclude<ExtType, 'yml'>;

export type Loader<T> = (path: Path, pkg: PackageStructure) => Promise<Partial<T>>;

export type Handler<T> = (prev: T, next: T) => Promise<T | undefined> | T | undefined;

export type FileSource = 'branch' | 'extended' | 'overridden' | 'root';

export interface File {
	/** Absolute path of the file. */
	path: Path;
	/** The source where the file originated. */
	source: FileSource;
}

export interface ConfigFile<T> extends File {
	/** Config content of the file. */
	config: DeepPartial<T>;
}

export interface IgnoreFile extends File {
	/** Ignored content of the file, split on new lines. */
	ignore: string[];
}

export interface ProcessedConfig<T> {
	/** All found and loaded config file contents merged and processed into a
  single config object. */
	config: Required<T>;
	/** List of config files found and loaded. */
	files: ConfigFile<T>[];
}

export interface ConfigFinderOptions<T> {
	/** Name of the setting in which "config extending" is enabled. */
	extendsSetting?: string;
	/** List of extensions, in order, to find config files within each folder. Defaults to built-in file format list. */
	extensions?: ExtType[];
	/** Find and load environment based config files (using `NODE_ENV`). Defaults to `true`. */
	includeEnv?: boolean;
	/** Mapping of loader functions by type. Defaults to built-in file type loaders. */
	loaders?: { [K in LoaderType]: Loader<T> };
	/** Name of config files, without extension. */
	name: string;
	/** Name of the setting in which "config overriding" is enabled. */
	overridesSetting?: string;
	/** Custom module resolver. */
	resolver?: ModuleResolver;
}

export interface IgnoreFinderOptions {
	/** Name of ignore files, without extension. */
	name: string;
}

export interface ProcessorOptions {
	/**
	 * When a setting has a value of `undefined`, fallback to the
	 * default/initial value for that setting. Defaults to `true`.
	 */
	defaultWhenUndefined?: boolean;
	/** Name of config files, without extension. */
	name: string;
	/** Validate all settings within a config file before processing. Defaults to
  `true`. */
	validate?: boolean;
}

export type FileGlob = string[] | string;

export type ExtendsSetting = string[] | string;

export interface OverridesSettingItem<T> {
	/** File path patterns/globs to ignore. */
	exclude?: FileGlob;
	/** File path patterns/globs to match against. */
	include: FileGlob;
	/** Settings configured for this specific override. */
	settings: DeepPartial<T>;
}

export type OverridesSetting<T> = OverridesSettingItem<T>[];

// PLUGINS

export type PluginOptions = boolean | object;

export type PluginEntry = string | [string, PluginOptions];

export type PluginsSettingMap = Record<string, PluginOptions>;

export type PluginsSettingList = PluginEntry[];

export type PluginsSetting = PluginsSettingList | PluginsSettingMap;
