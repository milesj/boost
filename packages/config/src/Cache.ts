import fs from 'fs';
import { Path } from '@boost/common';

export interface FileCache<T> {
	content: T;
	exists: boolean;
	mtime: number;
}

export default class Cache {
	configDir?: Path;

	dirFilesCache: Record<string, Path[]> = {};

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	fileContentCache: Record<string, FileCache<any>> = {};

	pkgPath?: Path;

	rootDir?: Path;

	async cacheFileContents<T>(path: Path, cb: () => Promise<T>): Promise<T> {
		const key = path.path();
		const cache = this.fileContentCache[key];
		const stats = await fs.promises.stat(path.path());

		if (cache && cache.mtime === stats.mtimeMs) {
			return cache.content;
		}

		const content = await cb();

		this.fileContentCache[key] = {
			content,
			exists: true,
			mtime: stats.mtimeMs,
		};

		return content;
	}

	async cacheFilesInDir(dir: Path, cb: () => Promise<Path[]>): Promise<Path[]> {
		const key = dir.path();

		if (this.dirFilesCache[key]) {
			return this.dirFilesCache[key];
		}

		const files = await cb();

		this.dirFilesCache[key] = files;

		return files;
	}

	clearFileCache() {
		this.fileContentCache = {};
	}

	clearFinderCache() {
		this.dirFilesCache = {};
	}

	getFileCache<T>(path: Path): FileCache<T> | null {
		return this.fileContentCache[path.path()] || null;
	}

	markMissingFile(path: Path): this {
		this.fileContentCache[path.path()] = {
			content: null,
			exists: false,
			mtime: 0,
		};

		return this;
	}
}
