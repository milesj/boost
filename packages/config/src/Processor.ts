import { isObject, Contract, Predicates } from '@boost/common';
import mergeArray from './mergeArray';
import mergeObject from './mergeObject';
import { Handler, ProcessorOptions } from './types';

export default class Processor<T extends object> extends Contract<ProcessorOptions> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected handlers: { [K in keyof T]?: Handler<any> } = {};

  blueprint({ string }: Predicates) {
    return {
      extendsSetting: string(),
      overridesSetting: string(),
    };
  }

  addHandler<K extends keyof T, V = T[K]>(key: K, handler: Handler<V>): this {
    this.handlers[key] = handler;

    return this;
  }

  process(defaults: Required<T>, configs: Partial<T>[]): Required<T> {
    return configs.reduce<Required<T>>((config, next) => this.runProcess(config, next), defaults);
  }

  protected runProcess(prev: Required<T>, next: Partial<T>): Required<T> {
    // const { extendsSetting, overridesSetting } = this.options;
    const config = { ...prev };

    Object.entries(next).forEach(([key, value]) => {
      const name = key as keyof T;
      const nextValue = value as T[keyof T];
      const prevValue = config[name];
      const handler = this.handlers[name];

      // Handlers only run on the root settings
      if (handler) {
        config[name] = handler(prevValue, nextValue);

        return;
      }

      if (isObject(prevValue) && isObject(nextValue)) {
        config[name] = mergeObject(prevValue, nextValue);
      } else if (Array.isArray(prevValue) && Array.isArray(nextValue)) {
        config[name] = mergeArray(prevValue, nextValue);
      } else if (nextValue !== undefined) {
        config[name] = nextValue;
      }
    });

    return config;
  }
}
