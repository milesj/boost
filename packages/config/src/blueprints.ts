import { predicates } from '@boost/common';
import { ExtendsSetting, PluginsSettingOption } from './types';

const { array, bool, object, string, union } = predicates;

export const extendsSetting = union<ExtendsSetting>(
  [string().notEmpty(), array(string().notEmpty())],
  [],
);

export const pluginsSetting = object(union<PluginsSettingOption>([bool(), object()], {}));
