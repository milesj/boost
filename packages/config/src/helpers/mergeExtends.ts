import { toArray } from '@boost/common';
import { ExtendsSetting } from '../types';
import mergeArray from './mergeArray';

export default function mergeExtends(prev: ExtendsSetting, next: ExtendsSetting): string[] {
	return mergeArray(toArray(prev), toArray(next));
}
