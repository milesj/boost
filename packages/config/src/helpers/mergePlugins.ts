import { isObject } from '@boost/common';
import { PluginsSetting, PluginsSettingList, PluginsSettingMap } from '../types';
import { mergeObject } from './mergeObject';

function convertListToMap(list: PluginsSettingList): PluginsSettingMap {
	return list.reduce((map, entry) => {
		const [name, options = true] = Array.isArray(entry) ? entry : [entry];

		return {
			...map,
			[name]: options,
		};
	}, {});
}

export function mergePlugins(prev: PluginsSetting, next: PluginsSetting): PluginsSettingMap {
	const plugins = Array.isArray(prev) ? convertListToMap(prev) : { ...prev };

	Object.entries(Array.isArray(next) ? convertListToMap(next) : next).forEach(([name, options]) => {
		if (isObject(options)) {
			plugins[name] = isObject(plugins[name])
				? mergeObject(plugins[name] as object, options)
				: options;
		} else if (options !== undefined) {
			plugins[name] = options;
		}
	});

	return plugins;
}
