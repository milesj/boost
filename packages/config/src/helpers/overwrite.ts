/**
 * Overwrite the previous value with the next value.
 */
export function overwrite<T>(prev: T, next: T): T {
	return next;
}
