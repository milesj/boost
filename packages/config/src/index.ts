/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './ConfigError';
export * from './Configuration';
export * from './helpers/getEnv';
export * from './helpers/mergeArray';
export * from './helpers/mergeExtends';
export * from './helpers/mergeObject';
export * from './helpers/mergePlugins';
export * from './helpers/overwrite';
export * from './schemas';
export * from './types';
