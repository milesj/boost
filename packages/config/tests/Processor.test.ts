import { Blueprint, predicates } from '@boost/common';
import { mockFilePath } from '@boost/common/test';
import { mergeExtends } from '../src/helpers/mergeExtends';
import { mergePlugins } from '../src/helpers/mergePlugins';
import { overwrite } from '../src/helpers/overwrite';
import {
	createExtendsPredicate,
	createOverridesPredicate,
	createPluginsPredicate,
} from '../src/predicates';
import { Processor } from '../src/Processor';
import { ConfigFile, ExtendsSetting, OverridesSetting, PluginsSetting } from '../src/types';

describe('Processor', () => {
	interface ConfigShape {
		// Settings
		debug: boolean;
		extends: ExtendsSetting;
		plugins: PluginsSetting;
		overrides: OverridesSetting<Omit<ConfigShape, 'extends' | 'overrides'>>;
		// Types
		boolean: boolean;
		string: string;
		stringList: string[];
		number: number;
		numberList: number[];
		object: object;
	}

	let processor: Processor<ConfigShape>;

	function stubConfigFile(config: Partial<ConfigShape>): ConfigFile<ConfigShape> {
		return {
			config,
			path: mockFilePath('.boost.js'),
			source: 'branch',
		};
	}

	beforeEach(() => {
		processor = new Processor({ name: 'boost' });
	});

	describe('handlers', () => {
		it('sets a handler', () => {
			const spy = jest.fn();
			processor.addHandler('debug', spy);

			expect(processor.getHandler('debug')).toBe(spy);
		});

		it('overwrites a handler of the same name', () => {
			const spy1 = jest.fn();
			processor.addHandler('debug', spy1);

			const spy2 = jest.fn();
			processor.addHandler('debug', spy2);

			expect(processor.getHandler('debug')).toBe(spy2);
		});

		it('returns null if not defined', () => {
			expect(processor.getHandler('debug')).toBeNull();
		});
	});

	describe('processing', () => {
		const commonBlueprint = {
			debug: predicates.bool(false),
			plugins: createPluginsPredicate(),
			boolean: predicates.bool(true),
			string: predicates.string(''),
			stringList: predicates.array(predicates.string(), ['foo']),
			number: predicates.number(123),
			numberList: predicates.array(predicates.number(), []),
			object: predicates.object(),
		};

		const blueprint: Blueprint<ConfigShape> = {
			...commonBlueprint,
			extends: createExtendsPredicate(),
			overrides: createOverridesPredicate(commonBlueprint),
		};

		const defaults: ConfigShape = {
			debug: false,
			extends: [],
			plugins: {},
			overrides: [],
			boolean: true,
			string: '',
			stringList: ['foo'],
			number: 123,
			numberList: [],
			object: {},
		};

		it('returns defaults if no configs', async () => {
			expect(await processor.process(defaults, [], blueprint)).toEqual(defaults);
		});

		describe('undefined values', () => {
			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ debug: true }), stubConfigFile({ debug: undefined })],
						blueprint,
					),
				).toEqual(defaults);
			});

			it('does not reset to default if `defaultWhenUndefined` option is false', async () => {
				processor.configure({ defaultWhenUndefined: false });

				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ debug: true }), stubConfigFile({ debug: undefined })],
						blueprint,
					),
				).toEqual({
					...defaults,
					debug: undefined,
				});
			});
		});

		describe('validation', () => {
			it('validates each config', async () => {
				await expect(
					processor.process(
						defaults,
						[
							stubConfigFile({ debug: true }),
							stubConfigFile({
								// @ts-expect-error Invalid type
								debug: 123,
							}),
						],
						blueprint,
					),
				).rejects.toThrowErrorMatchingSnapshot();
			});

			it('doesnt validate if the `validate` option is false', async () => {
				processor.configure({ validate: false });

				await expect(
					processor.process(
						defaults,
						[
							stubConfigFile({ debug: true }),
							stubConfigFile({
								// @ts-expect-error Invalid type
								debug: 123,
							}),
						],
						blueprint,
					),
				).resolves.not.toThrow();
			});
		});

		describe('extends setting', () => {
			it('supports extends from multiple configs', async () => {
				processor.addHandler('extends', mergeExtends);

				expect(
					await processor.process(
						defaults,
						[
							stubConfigFile({ extends: 'foo' }),
							stubConfigFile({ extends: ['bar', './baz.js'] }),
							stubConfigFile({ extends: '/qux.js' }),
						],
						blueprint,
					),
				).toEqual({
					...defaults,
					extends: ['foo', 'bar', './baz.js', '/qux.js'],
				});
			});

			it('errors for an invalid value', async () => {
				await expect(
					processor.process(
						defaults,
						[
							stubConfigFile({
								// @ts-expect-error Invalid type
								extends: 123,
							}),
						],
						blueprint,
					),
				).rejects.toThrowErrorMatchingSnapshot();
			});

			it('errors for an empty string', async () => {
				await expect(
					processor.process(defaults, [stubConfigFile({ extends: '' })], blueprint),
				).rejects.toThrowErrorMatchingSnapshot();
			});

			it('errors for an empty string in an array', async () => {
				await expect(
					processor.process(defaults, [stubConfigFile({ extends: ['foo', ''] })], blueprint),
				).rejects.toThrowErrorMatchingSnapshot();
			});
		});

		describe('plugins setting', () => {
			it('supports plugins from multiple configs', async () => {
				processor.addHandler('plugins', mergePlugins);

				expect(
					await processor.process(
						defaults,
						[
							stubConfigFile({ plugins: { foo: true, qux: { on: true } } }),
							stubConfigFile({ plugins: [['bar', false]] }),
							stubConfigFile({ plugins: { foo: { legacy: true }, baz: true, qux: { on: false } } }),
							stubConfigFile({ plugins: ['oop'] }),
						],
						blueprint,
					),
				).toEqual({
					...defaults,
					plugins: {
						foo: { legacy: true },
						bar: false,
						baz: true,
						qux: { on: false },
						oop: true,
					},
				});
			});

			it('errors for an invalid value', async () => {
				await expect(
					processor.process(
						defaults,
						[
							stubConfigFile({
								// @ts-expect-error Invalid type
								plugins: true,
							}),
						],
						blueprint,
					),
				).rejects.toThrowErrorMatchingSnapshot();
			});

			it('errors for an invalid option value', async () => {
				await expect(
					processor.process(
						defaults,
						[
							stubConfigFile({
								// @ts-expect-error Invalid type
								plugins: {
									foo: 123,
								},
							}),
						],
						blueprint,
					),
				).rejects.toThrowErrorMatchingSnapshot();
			});
		});

		describe('booleans', () => {
			it('can overwrite', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ debug: true }), stubConfigFile({ boolean: false })],
						blueprint,
					),
				).toEqual({
					...defaults,
					debug: true,
					boolean: false,
				});
			});

			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ debug: true }), stubConfigFile({ debug: undefined })],
						blueprint,
					),
				).toEqual(defaults);
			});
		});

		describe('strings', () => {
			it('can overwrite', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ string: 'foo' }), stubConfigFile({ string: 'bar' })],
						blueprint,
					),
				).toEqual({
					...defaults,
					string: 'bar',
				});
			});

			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ string: 'foo' }), stubConfigFile({ string: undefined })],
						blueprint,
					),
				).toEqual(defaults);
			});
		});

		describe('numbers', () => {
			it('can overwrite', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ number: 123 }), stubConfigFile({ number: 456 })],
						blueprint,
					),
				).toEqual({
					...defaults,
					number: 456,
				});
			});

			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ number: 999 }), stubConfigFile({ number: undefined })],
						blueprint,
					),
				).toEqual(defaults);
			});
		});

		describe('arrays', () => {
			it('merges and flattens by default', async () => {
				expect(
					await processor.process(
						defaults,
						[
							stubConfigFile({ stringList: ['foo', 'bar'] }),
							stubConfigFile({ stringList: [] }),
							stubConfigFile({ stringList: ['bar', 'baz'] }),
						],
						blueprint,
					),
				).toEqual({
					...defaults,
					stringList: ['foo', 'bar', 'baz'],
				});
			});

			it('can overwrite using a custom handler', async () => {
				processor.addHandler('numberList', overwrite);

				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ numberList: [1, 2, 3] }), stubConfigFile({ numberList: [4, 5, 6] })],
						blueprint,
					),
				).toEqual({
					...defaults,
					numberList: [4, 5, 6],
				});
			});

			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[
							stubConfigFile({ stringList: ['foo', 'bar', 'baz'] }),
							stubConfigFile({ stringList: undefined }),
						],
						blueprint,
					),
				).toEqual(defaults);
			});
		});

		describe('objects', () => {
			it('shallow merges by default', async () => {
				expect(
					await processor.process(
						defaults,
						[
							stubConfigFile({ object: { foo: 123, nested: { a: true }, baz: false } }),
							stubConfigFile({ object: {} }),
							stubConfigFile({ object: { foo: 456, nested: { b: true }, bar: 'abc' } }),
						],
						blueprint,
					),
				).toEqual({
					...defaults,
					object: { foo: 456, nested: { b: true }, bar: 'abc', baz: false },
				});
			});

			it('can overwrite using a custom handler', async () => {
				processor.addHandler('object', overwrite);

				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ object: { foo: 123 } }), stubConfigFile({ object: { bar: 456 } })],
						blueprint,
					),
				).toEqual({
					...defaults,
					object: { bar: 456 },
				});
			});

			it('can reset to default using undefined', async () => {
				expect(
					await processor.process(
						defaults,
						[stubConfigFile({ object: { bar: 456 } }), stubConfigFile({ object: undefined })],
						blueprint,
					),
				).toEqual(defaults);
			});
		});
	});
});
