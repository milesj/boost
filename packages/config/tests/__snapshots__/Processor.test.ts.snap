// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Processor > processing > extends setting > errors for an empty string 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid field "extends". String cannot be empty.]
`;

exports[`Processor > processing > extends setting > errors for an empty string in an array 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid member "extends[1]". String cannot be empty.]
`;

exports[`Processor > processing > extends setting > errors for an invalid value 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid field "extends". Received number but value must be one of: string, array<string>.]
`;

exports[`Processor > processing > plugins setting > errors for an invalid option value 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid field "plugins.foo". Received number but value must be one of: boolean, object.]
`;

exports[`Processor > processing > plugins setting > errors for an invalid value 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid field "plugins". Received boolean but value must be one of: array<string | tuple<string, boolean | object>>, object<boolean | object>.]
`;

exports[`Processor > processing > validation > validates each config 1`] = `
[OptimalError: The following validations have failed for \`.boost.js\`:
  - Invalid field "debug". Must be a boolean, received number.]
`;

exports[`Processor processing extends setting errors for an empty string 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid field "extends". String cannot be empty."
`;

exports[`Processor processing extends setting errors for an empty string in an array 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid member "extends[1]". String cannot be empty."
`;

exports[`Processor processing extends setting errors for an invalid value 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid field "extends". Received number but value must be one of: string, array<string>."
`;

exports[`Processor processing plugins setting errors for an invalid option value 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid field "plugins.foo". Received number but value must be one of: boolean, object."
`;

exports[`Processor processing plugins setting errors for an invalid value 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid field "plugins". Received boolean but value must be one of: array<string | tuple<string, boolean | object>>, object<boolean | object>."
`;

exports[`Processor processing validation validates each config 1`] = `
"The following validations have failed for \`.boost.js\`:
  - Invalid field "debug". Must be a boolean, received number."
`;
